### DISCUSSION TITLE
Inverse contracts

### DISCUSSION THREAD

**Discussion by alenpavlovich** [kernc/backtesting.py#414] [Q&A]
Hi team, 

How complicated would be to add inverse contracts in the backtesting library? PnL in this case is calculated differently.

Long position Profit & Loss = Contract Value x (1/Entry Price - 1/Exit Price)

More details: 
https://help.bybit.com/hc/en-us/articles/360039261054-What-is-Inverse-Contract-

PS. I am not associated with Bybit in any way, just for reference.

**Comment by kernc** [comment]
Isn't this just a mathematical inverse, i.e.:
```py
inverse_ohlc_df = -1 / df

bt = Backtest(inverse_ohlc_df, ...)
```

**Reply by alenpavlovich** [reply]
Well, not really. You're still trying to make your trading decision on the original price. 
The inverse contract will give you exposure to the price movement of BTC in USD term while its settlement is in BTC, hence the contract is inverse, and PnL is calculated as 1/ entry - 1/exit.

So your trading logic will never be derived from the inverse prices. You just need to calculate PnL differently.

**Reply by kernc** [reply]
> Well, not really. You're still trying to make your trading decision on the original price.

How about this:
```py
inverse_ohlc_df = (-1 / df).join(df.rename(columns='orig_{}'.format))

bt = Backtest(inverse_ohlc_df, ...)
```
Now the framework will use inverse prices (_1/entry - 1/exit_) for its internal state, while you can make your trading decision on the original price in columns "orig_Open", "orig_High", "orig_Low", "orig_Close" ...

**Reply by alenpavlovich** [reply]
In my view this should be separate property of the class. 

```py
    def pl(self):
        """Trade profit (positive) or loss (negative) in cash units."""
        price = self.__exit_price or self.__broker.last_price
        return self.__size * (price - self.__entry_price)
```
        
        
You're having exit-entry price here, which is correct for linear contracts but not for inverse contracts. The inverse is 1/entry - 1/exit. So they will have a wrong sign of the PnL.

**Reply by alenpavlovich** [reply]
If I would like to implement an inverse contract, any suggestion or advice on where the update will be required?

Need another function for sure 
def pl_inv(self)

**Reply by alenpavlovich** [reply]
Any thoughts?

**Reply by kernc** [reply]
> You're having exit-entry price here, which is correct for linear contracts but not for inverse contracts. The inverse is 1/entry - 1/exit. So they will have a wrong sign of the PnL.

Ah! :bulb: I think an additional [sign change](https://github.com/kernc/backtesting.py/discussions/414#discussioncomment-1011264) therefore is required. Can you confirm?

**Reply by alenpavlovich** [reply]
def pl(self):
        """Trade profit (positive) or loss (negative) in cash units."""
        price = self.__exit_price or self.__broker.last_price
        return self.__size * ((1/self.__entry_price) - (1/ price))
        
I haven't tested this yet, trying to see in the code where the changes needs to be done. If you can help on that end that would be great. 


Thinking to add it here:

class Backtest:
    """
    Backtest a particular (parameterized) strategy
    on particular data.

    Upon initialization, call method
    `backtesting.backtesting.Backtest.run` to run a backtest
    instance, or `backtesting.backtesting.Backtest.optimize` to
    optimize it.
    """
    def __init__(self,
                 data: pd.DataFrame,
                 strategy: Type[Strategy],
                 *,
                 cash: float = 10_000,
                 commission: float = .0,
                 margin: float = 1.,
                 trade_on_close=False,
                 hedging=False,
                 **inverse=False,**
                 exclusive_orders=False
                 ):       


And then you can combine the pnl function and the inverse one based on the inverse argument.

**Reply by kernc** [reply]
Unfortunately, _no_. I'm opposed to introducing new API when a [workaround](https://github.com/kernc/backtesting.py/discussions/414#discussioncomment-1011264) (note the change of the sign) as simple as the suggested exists.

**Reply by alenpavlovich** [reply]
That's okay. I am updating the library for myself, so if you want to highlight where the changes need to be made I am happy to do it. 

I don't need to commit the changes or anything.
